examples/merge.c

```diff
@@ -0,0 +1,387 @@
+
+	if (is_unborn) {
+		const char *symbolic_ref;
+		git_reference *head_ref;
+
+		/* HEAD reference is unborn, lookup manually so we don't try to resolve it */
+		err = git_reference_lookup(&head_ref, repo, "HEAD");
+		if (err != 0) {
+			fprintf(stderr, "failed to lookup HEAD ref\n");
+			return -1;
+		}
+
+		/* Grab the reference HEAD should be pointing to */
+		symbolic_ref = git_reference_symbolic_target(head_ref);
+
+		/* Force-create that reference on the target OID */
+		err = git_reference_create(&target_ref, repo, symbolic_ref, target_oid, 1, NULL);
+		if (err != 0) {
+			fprintf(stderr, "failed to create master reference\n");
+			return -1;
+		}
+
+		git_reference_free(head_ref);
+	} else {
+		/* HEAD exists, just lookup and resolve */
+		err = git_repository_head(&target_ref, repo);
+		if (err != 0) {
+			fprintf(stderr, "failed to get HEAD reference\n");
+			return -1;
+		}
+	}
+
+	/* Lookup the target object */
+	err = git_object_lookup(&target, repo, target_oid, GIT_OBJ_COMMIT);
+	if (err != 0) {
+		fprintf(stderr, "failed to lookup OID %s\n", git_oid_tostr_s(target_oid));
+		return -1;
+	}
+
+	/* Force-checkout the result so the workdir is in the expected state */
+	ff_checkout_options.checkout_strategy = GIT_CHECKOUT_FORCE;
+	err = git_checkout_tree(repo, target, &ff_checkout_options);
+	if (err != 0) {
+		fprintf(stderr, "failed to checkout HEAD reference\n");
+		return -1;
+	}
+
+	/* Move the target reference to the target OID */
+	err = git_reference_set_target(&new_target_ref, target_ref, target_oid, NULL);
+	if (err != 0) {
+		fprintf(stderr, "failed to move HEAD reference\n");
+		return -1;
+	}
+
+	git_reference_free(target_ref);
+	git_reference_free(new_target_ref);
+	git_object_free(target);
+
+	return 0;
+}
+
+static int analyze_merge(git_repository *repo, merge_options *opts)
+{
+	git_merge_options merge_opts = GIT_MERGE_OPTIONS_INIT;
+	git_checkout_options checkout_opts = GIT_CHECKOUT_OPTIONS_INIT;
+	git_merge_analysis_t analysis;
+	git_merge_preference_t preference;
+	int err = 0;
+
+	merge_opts.flags = 0; // GIT_MERGE_FIND_RENAMES;
+	merge_opts.file_flags = GIT_MERGE_FILE_STYLE_DIFF3;
+
+	checkout_opts.checkout_strategy = GIT_CHECKOUT_FORCE|GIT_CHECKOUT_ALLOW_CONFLICTS;
+
+	err = git_merge_analysis(&analysis, &preference,
+	                         repo,
+	                         (const git_annotated_commit **)opts->annotated,
+	                         opts->annotated_count);
+	check_lg2(err, "merge analysis failed", NULL);
+
+	if (analysis & GIT_MERGE_ANALYSIS_UP_TO_DATE) {
+		printf("Already up-to-date\n");
+		return 0;
+	} else if (analysis & GIT_MERGE_ANALYSIS_UNBORN ||
+		       (analysis & GIT_MERGE_ANALYSIS_FASTFORWARD &&
+		       !(preference & GIT_MERGE_PREFERENCE_NO_FASTFORWARD))) {
+		const git_oid *target_oid;
+		if (analysis & GIT_MERGE_ANALYSIS_UNBORN) {
+			printf("Unborn\n");
+		} else {
+			printf("Fast-forward\n");
+		}
+
+		/* Since this is a fast-forward, there can be only one merge head */
```

Comment:
s/can/should. You should probably make sure that's actually the case.

Reply:
Technically, the merge analysis would not have said that in that case (that would be an octopus merge). I made it an assert, just in case. *Unless* you can octopus-merge unborn branches, maybe 😉 ?

Reply:
Fair enough. Just checked `git_merge_analysis`, and the first thing we do is make sure that `their_heads_len == 1`.

Reply:
Huh. Doesn't this even imply that we are unable to do octopus merges at all? I honestly don't know right now. In that case, the code might even be greatly simplified

Reply:
[Yep](https://github.com/libgit2/libgit2/blob/master/src/merge.c#L3264). I'm just aiming for completeness though and mirroring the `git-merge` tool.